components/Init/src/virtio_con.c

```diff
@@ -0,0 +1,227 @@
+ *
+ * @todo Handling the RX and TX is similar between the x86/ARM VMMs, hence this file
+ * and `components/VM_Arm/src/modules/virtio_con.c` share a lot of common code.
+ * A refactor for the driver layer is needed.
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_con.h>
+#include <sel4vmmplatsupport/device.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <platsupport/serial.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include "virtio_con.h"
+#include "virtio_irq.h"
+
+static virtio_con_t *virtio_con;
+
+/* Used as the parameter for callback functions (currently only `console_handle_irq`) */
+typedef struct virtio_con_cookie {
+    virtio_con_t *virtio_con;
+    vm_t *vm;
+} virtio_con_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(serial_layout)
+
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE (0x1000 - 2 * sizeof(uint32_t))
+
+/**
+ * This struct is a ring buffer that occupies exactly one page, it should represent
+ * the data in the shmem region (that is shared between two nodes of a virtio console
+ * connection.
+ *
+ * The access can be lock-free since an instance of this struct can only be either the
+ * receive buffer or the sent buffer of a virtio console node, which guarantees that only
+ * one end of the ring buffer struct can be modified in a node.
+
+ * Way of indexing: `head` and `tail` should not increase beyond (BUFSIZE - 1), as
+ * they represent the real index in `buf`.
+ */
+typedef struct serial_shmem {
+    uint32_t head;
+    uint32_t tail;
+    char buf[BUFSIZE];
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
+
+/**
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct serial_conn {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
+} serial_conn_t;
+static serial_conn_t connections[NUM_PORTS];
+
+/* defined in virtio_console_emul.c */
+void virtio_console_putchar(int port, virtio_emul_t *con, char *buffer, uint32_t head, uint32_t tail) WEAK;
+
+/**
+ * Writes data to the guest console attached to a port, and sets the head of the receive
+ * ring buffer to the right place.
+ *
+ * @param port Port number
+ */
+static void virtio_con_notify_recv(int port)
+{
+    volatile serial_shmem_t *buffer = connections[port].recv_buf;
+    virtio_console_putchar(port, virtio_con->emul, (char *)buffer->buf, buffer->head, buffer->tail);
+    buffer->head = buffer->tail;
+}
+
+/**
+ * Callback function for camkes virtqueue notification. Polls each receive ring buffer
+ * to see if there are data pending.
+ *
+ * @see main.c (struct device_notify) for parameters
+ */
+void handle_serial_console(vm_t *vmm)
```

Comment:
Maybe add a comment that for some legacy reason we get a `vmm` here, that we don't use. Where does it come from actually, should we have an assert that our connection VM has the same VMM than the VMM passed here?

Reply:
Can't think of the reason why the vm handler is there in the first place, since none of the implementations of this callback makes use of it. Maybe it was there for debugging purposes? I suggested the parameter should be removed in future work.